src/machine/machined-dbus.c

```diff
@@ -321,6 +322,15 @@ static int method_create_or_register_machine(
         if (r < 0)
                 return r;
 
+        /* Ensure an unprivileged user cannot claim any process they don't control as their own machine */
+        if (uid != 0) {
+                r = process_is_owned_by_uid(&leader_pidref, uid);
```

Comment:
If I see it right then this works in the Varlink case, because of `machine_pidref()` in combination with `SD_JSON_STRICT` (it requires the client to pass a pidfd).

In the D-Bus case here I believe there's a loophole above in line 298:

```
r = pidref_set_pid(&leader_pidref, leader);
```

This will cause the pidfd to opened in the daemon from the PID provided by the client and thus suffers from a race condition.

Reply:
Yeah good catch, I'll add new methods that take pidfds as params

Reply:
is this really an issue? i mean, sure, we might end up with the wrong process, but it's not *any* process, it's just some *other* process owned by the *same* user. And we early on pin the rpocess via a pidfd on machined's side, hence this check is definitely against the same process we eventually end up using as leader for the machine, hence we should be fine here, no?

i.e. passing us a simple pid is always racy, no doubt. but it's not necessarily a security issue as long as the wrong process we end up using is from the same security domain (i.e. same uid) as the intended one.

Reply:
I guess it's better safe than sorry. I'll open a separate PR for that, given it's way too late for new dbus interfaces. But it would be nice to use end-to-end pidfds anyway.